Edit DatabaseManager.java to improve try blocks and documentation

mewilker opened the pull request:
This PR fixes issues with documentation and JDBC connections that were left open in the code. These issues were discussed in a TA meeting. When reviewing, please make sure the style of the try-with-resources is correctly implemented. 

In addition to this, I made two changes that were not discussed in the meeting. The first was making the createDatabase method public since most students end up using this function in their Server class and making it public anyway. 

The second is a code comment in the getConnection method warning students not to use a try-with-resources _inside_ this method or the whole function will return a closed connection--which would not work.

leesjensen reviewed (approved):
Looks great!

mewilker committed:
edit DatabaseManager.java to remove DBInfo

mewilker committed:
remove excessive external try from code that loads db.properties

mewilker committed:
edit DatabaseManager.createDatabase to use a try with resources on it's connection

mewilker committed:
edit DatabaseManager.createDatabase to be public so students by default can use it in their Server class

mewilker committed:
added a code comment in DatabaseManager.getConnection warning students not to wrap the Driver's getConnection with a try with resources

frozenfrank commented on starter-code/4-database/dataaccess/DatabaseManager.java:
```diff
@@ -55,13 +51,14 @@ static void createDatabase() throws DataAccessException {
      */
     static Connection getConnection() throws DataAccessException {
         try {
+            //wrapping this line with a try-with-resources block will return a closed connection
```
Thank you for including this helpful comment! I worked with a student who made this mistake, and it was really tricky to figure out. This is the perfect kind of comment to that describes _why_ this line of code will be different from the rest of the similar lines.

frozenfrank commented on starter-code/4-database/dataaccess/DatabaseManager.java:
```diff
@@ -36,13 +34,11 @@ public class DatabaseManager {
-            var statement = "CREATE DATABASE IF NOT EXISTS " + DATABASE_NAME;
-            var conn = DriverManager.getConnection(CONNECTION_URL, USER, PASSWORD);
-            try (var preparedStatement = conn.prepareStatement(statement)) {
+    static public void createDatabase() throws DataAccessException {
```
I support this change.

It honestly makes the most sense to call this method from the `Server` class, but this file is intended to be located in the `dataaccess` package. Therefore, making `createDatabase()` public but leaving `getConnection()` as package-private makes the most sense by default.

frozenfrank reviewed (approved):
Thanks for making these helpful improvements! Very good.

mewilker committed:
remove getConnection parameter that is not required

mewilker committed:
rephrase the comment so it's more clear that this is the only place to not use a try-with-resources

mewilker committed:
fix spacing DatabaseManager.createDatabase to match Intellij suggestions

jerodw reviewed (changes requested):
I like this change except for the changes to the throws clauses. We should keep those as "throw new DataAcessException(e)" so you can see the actual cause of the issue when debugging. This still follows the principle of hiding the fact that it's an SQLException because you aren't required to write catch blocks or throws clauses for SQLException in the callers so you aren't creating a dependency on SQLException

frozenfrank committed:
Improve exception throwing in DatabaseManager

This requires replacing the constructor in DataAccessException.

This has the unintended side-effect of also making it harder for
students to unintentionally use this exception for their own
custom exceptions where they should really be creating their
new custom exception classes.

frozenfrank committed:
Standardize error throwing...

All changes are in DatabaseManager, but all other starter-code files
were checked for consistency in the process. DatabaseManager is the
only starter code file that performs any explicit error catching/throwing.

Standardizations:
- Always name the exception variable 'ex'
- Use lowercase in error message
- Always include the exception in the wrapped throwable
  This also applies to the RuntimeException
- Adjust error handling string management to better handle
  the entire field of potential errors that could be thrown

frozenfrank commented:
> I like this change except for the changes to the throws clauses. We should keep those as "throw new DataAcessException(e)" so you can see the actual cause of the issue when debugging. This still follows the principle of hiding the fact that it's an SQLException because you aren't required to write catch blocks or throws clauses for SQLException in the callers so you aren't creating a dependency on SQLException

I agree with you that including the original exception would be better.

## Defense of PR Contributions

In defense of this PR, those lines weren't changed here. The existing code already used the form `throw new DataAccessException(e.getMessage())`.

Here are the two relevant code samples, as they currently exist on the `main` branch:

https://github.com/softwareconstruction240/chess/blob/b3d94c165012348ecf7abf2a7abd71479e3ce7b7/starter-code/4-database/dataaccess/DatabaseManager.java#L39-L49

https://github.com/softwareconstruction240/chess/blob/b3d94c165012348ecf7abf2a7abd71479e3ce7b7/starter-code/4-database/dataaccess/DatabaseManager.java#L63-L71


In fact, the `DataAccessException` doesn't yet have a constructor to accept this more detailed format. See the current source:

https://github.com/softwareconstruction240/chess/blob/b3d94c165012348ecf7abf2a7abd71479e3ce7b7/starter-code/3-web-api/dataaccess/DataAccessException.java#L6-L10

## Response to Changes Requested

I will add a new constructor to DataAccessException in order to accommodate this request.

frozenfrank commented:
Done ✅ 

The error throwing has been standardized and improved in commit 4bf42d6cc04de29fc92931bb3c8444ebd9b5f4b4:

    Improve exception throwing in DatabaseManager
    
    This requires replacing the constructor in DataAccessException.
    
    This has the additional side-effect of also making it harder for
    students to unintentionally use this exception for their own
    custom exceptions where they should really be creating their
    new custom exception classes.

jerodw commented:
I compared it to my version and assumed it was changed, but I guess I changed it in my own implementation. I like the new constructor that allows including the original exception, but I didn't intend to make it impossible to create a DataAccessException with just a String. I just intended for our uses within DatabaseManager to include the original exception. Please add an additional constructor that takes just a String.

frozenfrank committed:
Update DataAccessException.java Restore String only constructor

This fulfills the request by @jerodw.

Currently, this new constructor is unused. 
@19mdavenport Will this cause any issues with the 
style checker and unused code?

mewilker committed:
specifically state the db.properties file not found in error message

frozenfrank committed:
Revert "specifically state the db.properties file not found in error message"

This reverts commit c6282dfec2e5a63e4b21fb2334ef3bdedc6b5589.

The intended benefit is already covered by the 'catch' clause
on line 30 which rethrows the exception (and other exceptions too)
with the specific name of the file.

frozenfrank commented:
@mewilker The most recent commit (c6282df) is not necessary. Notice that the thrown error is immediately caught by the catch block, and then rethrown on line 30. The encapsulating `RuntimeException` includes the specific file name and the original error. However, it is more general to include the file name for any other kind of error that could also be thrown.

https://github.com/softwareconstruction240/chess/blob/48f17f56ff09ba0eb61de402b322fdc480dda940/starter-code/4-database/dataaccess/DatabaseManager.java#L16-L31
